### PR TITLE
Secure upgrade HLD

### PR DESCRIPTION
**Pull request by ycoheNvidia**
This PR contains the Secure Upgrade HLD.

| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [[sonic-buildimage] Added support for Secure Upgrade](https://github.com/Azure/sonic-buildimage/pull/11862) | GitHub issue/pull request detail | GitHub pull request check contexts |
| [[sonic-utilities] Added support for Secure Upgrade](https://github.com/Azure/sonic-utilities/pull/2337) | GitHub issue/pull request detail | GitHub pull request check contexts |

### CODE REVIEW

**Comment by zhangyanzhao** [comment]
@madhupalu from Aviz Networks will help to review this.

**Comment by liat-grozovik** [comment]
@yxieca , @madhupalu as stated in the 202211 release features list you are the reviewers. Could you please help to review?
Nvidia will soon have the code PRs avaialble thus need to close the HLD first.

**Comment by ycoheNvidia** [comment]
| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [[sonic-buildimage] Added support for Secure Upgrade]([https://github.com/Azure/sonic-buildimage/pull/11862](https://github.com/Azure/sonic-buildimage/pull/10323)) | GitHub issue/pull request detail) | GitHub pull request check contexts) |
| [[sonic-utilities Added support for Secure Upgrade](https://github.com/Azure/sonic-utilities/pull/2337) | GitHub issue/pull request detail) | GitHub pull request check contexts |

**Comment by liat-grozovik** [comment]
@yxieca , @madhupalu kindly reminder for review comments.

**Comment by yxieca** [comment]
@vaibhavhd  to check from upgrade perspective. @xumia to check from secure boot/upgrade perspective.

File: doc/secure_upgrade/secure_upgrade.md
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:65]: correct typo "flowing" here - is the secure upgrade verification commit ready? if yes please share it

File: doc/secure_upgrade/secure_upgrade.md
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:102]: correct BOIS type

File: doc/secure_upgrade/secure_upgrade.md
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:130]: what are the UEFI tools being used to support secure upgrade process?

File: doc/secure_upgrade/secure_upgrade.md
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:133]: Is the ONIE pull request is ready? please share if it is

File: doc/secure_upgrade/secure_upgrade.md
Comment (madhupalu) [inline doc/secure_upgrade/secure_upgrade.md:86]: How is the signing prod process is different from dev? How do you secure production keys? do you use vault kind of infra?

File: doc/secure_upgrade/secure_upgrade.md
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:65]: fix the typo, and for the moment the "commit" is waiting for review, added the current PR in the HLD

File: doc/secure_upgrade/secure_upgrade.md
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:102]: done

File: doc/secure_upgrade/secure_upgrade.md
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:130]: added the list and url

File: doc/secure_upgrade/secure_upgrade.md
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:133]: still not

File: doc/secure_upgrade/secure_upgrade.md
Comment (davidpil2002) [reply doc/secure_upgrade/secure_upgrade.md:86]: The feature is including the code that supports secure upgrade using development keys. And any Vendor should decide his own design flow for production. 
In general, in the build flow when the user is building the image with "prod" flag (as explained in chapter 1.7.1. Signing process) the build flow will call a signing script that should be coded by the vendor, this script should send the image to a different server "sign server" and this server should return the image signed.

**Comment by Yarden-Z** [comment]
@yxieca , @xumia can you please approve this?
We fixed the latest fault we had here and answered all the review questions.

**Comment by liat-grozovik** [comment]
This is the last call for the HLD review. Kindly reminder for those who provided feedback or still need to provide to have them available this week
@zhangyanzhao as this is for 202211 we need to close the loop. Also code PR is available for review

**Comment by Yarden-Z** [comment]
@xumia / @zhangyanzhao Is there any update on this HLD review?

**Comment by zhangyanzhao** [comment]
@yxieca  @madhupalu can you please help to review this HLD and related code PRs? Need your approval after your review. Thanks.

File: doc/secure_upgrade/secure_upgrade.md
Comment (xumia) [reply doc/secure_upgrade/secure_upgrade.md:86]: @davidpil2002 , we have a different environment for build production and dev, production is always built in security environment. Is it possible to use the same process for both production and dev? The only different is the certificates used, but the build steps are all the same.
It is good to provide an additional script to sign images from the "sign server", but can we support to the simple scenario to support "sign server" and build on the same machine? How about removing the "_DEV_" in the parameters?

Comment (ycoheNvidia) [reply doc/secure_upgrade/secure_upgrade.md:86]: @xumia thank you for the question. In many cases, different customers have different production servers with different environments. Therefor no single script will be applicable here. In the case you suggest - it is possible to copy the dev key as an argument to prod script, but this is a specific use case

**Review by xumia — APPROVED** [review]

**Comment by liat-grozovik** [comment]
The feature and its current content is expected to be in 202305 and into 202211 as originally aimed.

**Pull request merged**
